### ISSUE TITLE
Proposal: Task.WhenAllSuccessfulOrAnyFailed

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#21617]
Task currently exposes two When helpers: WhenAll and WhenAny.  These are the most commonly needed combinators, but there's another that's quite helpful, especially in tests: a combinatory that asynchronously waits for either all of the tasks to complete successfully or for the first one that doesn't complete successfully.  This is particularly valuable when writing tests that involve related asynchronous operations, such as in networking, for example where you want to have an asynchronous client operation (e.g. socket.ConnectAsync) and an asynchronous server operation (e.g. socket.AcceptAsync) that are directly dependent on each other... using WhenAll can lead to deadlocks if one of those operations fails causing the other to wait forever.

Proposal:
```C#
public class Task
{
    public static Task WhenAllSuccessfulOrAnyFailed(params Task[] tasks);
    public static Task WhenAllSuccessfulOrAnyFailed(IEnumerable<Task> tasks);
    ...
}
```

From a behavioral standpoint, it would be as if code like this were used:
```C#
var remaining = new List<Task>(tasks);
while (remaining.Count > 0)
{
    Task t = await Task.WhenAny(remaining).ConfigureAwait(false);
    remaining.Remove(t);
    t.GetAwaiter().GetResult();
}
```
but with a more efficient implementation.

I'm not sure I love the name either, but I'm sure we can come up with something.

**Comment by terrajobst** [comment]
@stephentoub wants to think about the naming and shape a bit more.

**Comment by egil** [comment]
I recently asked about something like this on [Twitter](https://twitter.com/egilhansen/status/1609892964590919683) and one of your colleagues pointed to an [existing implementation in this repository](https://github.com/dotnet/runtime/blob/2619d1c8eeef4a881c3910c87c1a8903ed742c24/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs#L60).

I like the name they use, `WhenAllOrAnyFailed`, since it nicely combines the name of the two existing helper methods that are already well-understood in the community. I do not think it's necessary to include the word "successful" as it seems implied by the "OrAnyFailed" part.

Here is my variation of implementation from this repo for reference:

```C#
public static Task WhenAllOrAnyFailed(params Task[] tasks)
{
    if (tasks.Length == 0)
    {
        return Task.CompletedTask;
    }

    var remaining = tasks.Length;
    var tcs = new TaskCompletionSource();

    foreach (var task in tasks)
    {
        task.ContinueWith(
            a =>
            {
                if (GetRealException(a) is Exception e)
                {
                    tcs.TrySetException(e);
                }
                else if (Interlocked.Decrement(ref remaining) == 0)
                {
                    tcs.TrySetResult();
                }
            },
            CancellationToken.None,
            TaskContinuationOptions.None,
            TaskScheduler.Default);
    }

    return tcs.Task;

    // Gets the exception (if any) from the Task, for both faulted and canceled tasks
    static Exception? GetRealException(Task task)
    {
        try
        {
            task.GetAwaiter().GetResult();
        }
        catch (Exception e)
        {
            return e;
        }

        return null;
    }
}
```

**Comment by theodorzoulias** [comment]
@egil be aware that in case one of the tasks fails, the `WhenAllOrAnyFailed` task will complete with failure without detaching the `ContinueWith` continuations that have been attached to the incomplete tasks. This can be a problem in case the `WhenAllOrAnyFailed` is called repeatedly in a loop for the same set of tasks. To fix this you could use a `CancellationTokenSource`, and cancel the continuations when a task fails.

**Comment by egil** [comment]
> @egil be aware that in case one of the tasks fails, the `WhenAllOrAnyFailed` task will complete with failure without detaching the `ContinueWith` continuations that have been attached to the incomplete tasks. This can be a problem in case the `WhenAllOrAnyFailed` is called repeatedly in a loop for the same set of tasks. To fix this you could use a `CancellationTokenSource`, and cancel the continuations when a task fails.

Good point. I actually do that in the code that calls it, but it's better to make that part of the `WhenAllOrAnyFailed` itself.
